[Lazyload] Refactor lazy load tests

domfarolino opened the pull request:
Prior to this PR, the tests existed as follows:

 - `image-loading-lazy.tentative.html`
   - This tested that `loading=lazy` images are not loaded until they are scrolled into the viewport, and incorrectly tested more-advanced load-blocking behavior; that is, it tested that in-viewport `loading=lazy` images would block the load event, which is not up to spec.
 - `invisible-lazy-image.tentative.html`
   - This tested that in-viewport `loading=lazy` images that should render (i.e., not `display:none`) _do not_ block the window load event, and that `loading=lazy` images that do not render (i.e., `display:none`) don't load at all.

After this PR, the tests exist as follows:
 - `image-loading-lazy.tentative.html`
   - This tests "basic" lazy load behavior/support, that is, it tests **1.)** that `loading=lazy` images are not loaded until they are scrolled into the viewport **2.)** that `loading=lazy` images that are below-viewport do not block the window load event from firing
 - `image-loading-lazy-load-event.tentative.html`
   - This tests more advanced lazy load behavior/support, that is, it specifically exercises the `delay load event` flag in the spec, by asserting that in-viewport images that should render (i.e., not `display:none`) do not block the window load event even when they start loading before the window load event fires. It contains a subset of the tests in the old `invisible-lazy-image.tentative.html` file.
 - `below-viewport-image-loading-lazy-load-event.tentative.html`
   - Like the above test, also exercises the `delay load event` spec flag. Does so by having below-viewport `loading=lazy` renderable (not `display:none`) images scrolled into view before the window load event fires, and asserts that these images start loading, but do not block the window load event from firing.
 - `not-rendered-image-loading-lazy.tentative.html`
   - This file is a port of the non-renderable-image tests that used to be in `invisible-lazy-image.tentative.html`. I've moved them here to break up the tests in the old file, so that it isn't too complicated.
 - `not-rendered-below-viewport-image-loading-lazy.tentative.html`
   - Like the above, tests that non-renderable below-viewport `loading=lazy` don't block the window load event, and never load even once scrolled into view. It contains much of the logic in the old `invisible-lazy-image.tentative.html`.

domfarolino committed:
Refactor lazy loading tests 1/n

domfarolino committed:
Undo multi changes

domfarolino committed:
Remove console.log

domfarolino committed:
Add not-rendered-image-loading-lazy test

domfarolino committed:
Add non-rendered image tests (in- and below-viewport)

domfarolino committed:
Fix taskcluster

domfarolino committed:
Small fixes, and abstract-out unreached_funcs

rwlbuis commented on loading/lazyload/not-rendered-below-viewport-image-loading-lazy.tentative.html:
```diff
@@ -0,0 +1,67 @@
+
+    const unreached_not_rendered_img_func =
+      t.unreached_func("The not-rendered below-viewport loading=lazy images " +
+                       "do not attempt to load.");
```
Is "should not" better here?

rwlbuis reviewed (commented):
Looks great overall!
Do we want to add more onerror handling for consistency?

domfarolino committed:
More error handling + assertion messages

domfarolino commented:
I added some more error handling for tests using ElementLoadPromise, and fixed any assertion messages in this change that https://github.com/web-platform-tests/wpt/pull/20783#discussion_r358123442 applies to. For tests that do not use ElementLoadPromise, it is a little clunkier adding error handling to them, so I suggest we separately consider converting them to use ElementLoadPromise and add such handling.

rwlbuis commented on loading/lazyload/image-loading-lazy.tentative.html:
```diff
@@ -1,55 +1,52 @@
+  <!-- |in_viewport| takes 2 seconds to load, so that in browsers that don't
+       support lazy loading, |below_viewport| finishes before |in_viewport|, and
+       the test will dependably fail without relying on a timeout. -->
+  <img id="in_viewport" loading="lazy" src="resources/image.png?first&pipe=trickle(d2)"
```
I guess this one needs onerror as well?

rwlbuis commented on loading/lazyload/image-loading-lazy.tentative.html:
```diff
@@ -1,55 +1,52 @@
+  <img id="in_viewport" loading="lazy" src="resources/image.png?first&pipe=trickle(d2)"
+       onload="in_viewport_img_onload();">
+  <div style="height:1000vh;"></div>
+  <img id="below_viewport" loading="lazy" src="resources/image.png?second"
```
Ditto?

rwlbuis replied:
Never mind.

domfarolino replied:
Yeah, maybe we can convert these to ElementLoadPromise for simplicity later

rwlbuis reviewed (approved):
LGTM!

domfarolino commented:
Thanks! Merging this now, but would also welcome @zcorpan's post-commit review whenever :)
